bin/proving-service/src/proxy/mod.rs

```diff
@@ -193,25 +191,23 @@ impl LoadBalancerState {
         self.workers.read().await.iter().filter(|w| !w.is_available()).count()
     }
 
-    /// Check the health of the workers and returns a list of healthy workers.
+    /// Check the health of the workers.
     ///
-    /// Performs a health check on each worker using the gRPC health check protocol. If a worker
-    /// is not healthy, it won't be included in the list of healthy workers.
-    async fn check_workers_health(
-        &self,
-        workers: impl Iterator<Item = &mut Worker>,
-    ) -> Vec<Worker> {
-        let mut healthy_workers = Vec::new();
-
+    /// Performs a health check on each worker using the gRPC health check protocol.
+    ///
+    /// If a worker is not healthy, it will be marked as unhealthy and the number of unhealthy
+    /// workers will be incremented.
+    async fn check_workers_health(&self, workers: impl Iterator<Item = &mut Worker>) {
         for worker in workers {
-            if worker.is_healthy().await {
-                healthy_workers.push(worker.clone());
-            } else {
-                warn!("Worker {} is not healthy", worker.address());
+            if let Some(is_healthy) = worker.is_healthy().await {
+                if is_healthy {
+                    worker.mark_as_healthy();
+                } else {
+                    worker.mark_as_unhealthy();
+                    WORKER_UNHEALTHY.with_label_values(&[&worker.address()]).inc();
+                }
```

Comment:
Note that slow workers will block here.

Ideally each worker would have its own independent health check routine/task.